homeassistant/helpers/service.py

```diff
@@ -783,3 +783,43 @@ async def check_permissions(call: ServiceCall) -> Any:
         return check_permissions
 
     return decorator
+
+
+class ReloadServiceHelper:
+    """Helper for reload services to minimize unnecessary reloads."""
+
+    def __init__(self, service_func: Callable[[ServiceCall], Awaitable]):
+        """Initialize ReloadServiceHelper."""
+        self._service_func = service_func
+        self._service_running = False
+        self._service_condition = asyncio.Condition()
+
+    async def execute_service(self, service_call: ServiceCall) -> None:
+        """Execute the service.
+
+        If a previous reload task if currently in progress, wait for it to finish first.
+        Once the previous reload task has finished, one of the waiting tasks will be
+        assigned to execute the reload, the others will wait for the reload to finish.
+        """
+
+        do_reload = False
+        async with self._service_condition:
+            if self._service_running:
+                # A previous reload task is already in progress, wait for it to finish
+                await self._service_condition.wait()
+
+        async with self._service_condition:
+            if not self._service_running:
+                # This task will do the reload
+                self._service_running = True
+                do_reload = True
+            else:
+                # Another task will perform the reload, wait for it to finish
+                await self._service_condition.wait()
+
+        if do_reload:
+            # Reload, then notify other tasks
+            await self._service_func(service_call)
+            async with self._service_condition:
```

Comment:
Wouldn't this wait until all service conditions are released while above there is code to keep this lock until they are notified. Seems like a deadlock.

Reply:
There is no deadlock, the lock is [released](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Condition.wait) above in the call to `await self._service_condition.wait()`
The point of the condition variable here is to guarantee exclusive access to the `self._service_condition` flag to remove any races when multiple tasks are concurrently accessing and potentially updating it.

Reply:
Ah, I missed that part of the docs. Ok all good.